wom/src/main/scala/wdl4s/wom/graph/CallNode.scala

```diff
@@ -54,29 +59,34 @@ object CallNode {
   /**
     * Create a CallNode for a Callable (task or workflow).
     *
-    * If an input is supplied, it gets wired in as appropriate.
+    * If an input is supplied as a port from another Node, it gets wired in directly.
+    * If an input is supplied as an expression, we try to create an InstantiatedExpression and include that in the call.
     * If an input is not supplied, it gets created as a GraphInputNode.
     *
-    * The returned value is a tuple of (
-    *   _1: the CallNode
-    *   _2: any GraphInputNodes we created for unsupplied inputs
-    * )
     */
-  def callWithInputs(name: String, callable: Callable, inputMapping: Map[String, OutputPort]): CallWithInputs = {
+  def callWithInputs(name: String, callable: Callable, portInputs: Map[String, OutputPort], expressionInputs: Set[GraphNodeInputExpression]): ErrorOr[CallWithInputs] = {
 
     val graphNodeSetter = new GraphNode.GraphNodeSetter()
-    val inputPortLinker = GraphNode.linkInputPort(callable.name + ".", inputMapping, graphNodeSetter.get) _
-    val linkedInputPortsAndGraphInputNodes = callable.inputs map inputPortLinker
 
-    val linkedInputPorts = linkedInputPortsAndGraphInputNodes.map(_.newInputPort)
-    val graphInputNodes = linkedInputPortsAndGraphInputNodes collect { case LinkedInputPort(_, Some(gin)) => gin }
+    val instantiatedExpressionInputsAttempt: ErrorOr[Map[String, InstantiatedExpression]] = expressionInputs.toList traverse { gnie => gnie.instantiateExpression(graphNodeSetter) } map { _.toMap }
 
-    val callNode = callable match {
-      case t: TaskDefinition => TaskCallNode(name, t, linkedInputPorts)
-      case w: WorkflowDefinition => WorkflowCallNode(name, w, linkedInputPorts)
-    }
+    instantiatedExpressionInputsAttempt map { instantiatedExpressionInputs =>
+      val inputPortLinker = GraphNode.linkInputPort(callable.name + ".", portInputs, graphNodeSetter.get) _
+
+      // Filter out the inputs we already have from expressions:
+      val asYetUnsuppliedInputs = callable.inputs.filterNot(inputDef => instantiatedExpressionInputs.contains(inputDef.name))
+      val linkedInputPortsAndGraphInputNodes = asYetUnsuppliedInputs map inputPortLinker
+
+      val linkedInputPorts = linkedInputPortsAndGraphInputNodes.map(_.newInputPort)
+      val graphInputNodes = linkedInputPortsAndGraphInputNodes collect { case LinkedInputPort(_, Some(gin)) => gin }
 
-    graphNodeSetter._graphNode = callNode
-    CallWithInputs(callNode, graphInputNodes)
+      val callNode = callable match {
+        case t: TaskDefinition => TaskCallNode(name, t, linkedInputPorts, instantiatedExpressionInputs)
+        case w: WorkflowDefinition => WorkflowCallNode(name, w, linkedInputPorts, instantiatedExpressionInputs)
```

Comment:
I like the fact that CallNode instances can only be constructed by coming via `callWithInputs`, since it makes sure everything is in order when creating the Node.

But, I will extract this logic out into a trait-level `apply`, which might be less offensive to you...